pageserver/src/walreceiver.rs

```diff
@@ -64,375 +80,1148 @@ thread_local! {
+                            {
+                                Ok(ControlFlow::Break(())) => {
+                                    break;
+                                }
+                                Ok(ControlFlow::Continue(())) => {}
+                                Err(e) => warn!("Error during wal receiver main thread step for timeline {new_id}: {e:#}"),
+                            }
+                        }
+                    }.instrument(info_span!("timeline", id = %new_id)));
+
+                    vacant_timeline_entry.insert(TimelineWalBrokerLoopHandles {
+                        broker_join_handle,
+                        cancellation_sender,
+                    });
+                }
+            }
         }
         None => {
-            let thread_id = thread_mgr::spawn(
-                ThreadKind::WalReceiver,
-                Some(tenantid),
-                Some(timelineid),
-                "WAL receiver thread",
-                false,
-                move || {
-                    IS_WAL_RECEIVER.with(|c| c.set(true));
-                    thread_main(conf, tenantid, timelineid);
-                    Ok(())
-                },
-            )?;
-
-            let receiver = WalReceiverEntry {
-                thread_id,
-                wal_producer_connstr: wal_producer_connstr.into(),
-                last_received_msg_lsn: None,
-                last_received_msg_ts: None,
-            };
-            receivers.insert((tenantid, timelineid), receiver);
-
-            // Update tenant state and start tenant threads, if they are not running yet.
-            tenant_mgr::activate_tenant(tenantid)?;
+            info!("Local timeline update channel closed");
+            shutdown_all_wal_connections(local_timeline_wal_receivers).await;
         }
-    };
-    Ok(())
+    }
 }
 
-/// Look up a WAL receiver's data in the global `WAL_RECEIVERS`
-pub fn get_wal_receiver_entry(
+async fn fetch_tenant_settings(
     tenant_id: ZTenantId,
-    timeline_id: ZTimelineId,
-) -> Option<WalReceiverEntry> {
-    let receivers = WAL_RECEIVERS.lock().unwrap();
-    receivers.get(&(tenant_id, timeline_id)).cloned()
+) -> anyhow::Result<(Duration, Duration, NonZeroU64)> {
+    tokio::task::spawn_blocking(move || {
+        let repo = tenant_mgr::get_repository_for_tenant(tenant_id)
+            .with_context(|| format!("no repository found for tenant {tenant_id}"))?;
+        Ok::<_, anyhow::Error>((
+            repo.get_wal_receiver_connect_timeout(),
+            repo.get_lagging_wal_timeout(),
+            repo.get_max_lsn_wal_lag(),
+        ))
+    })
+    .await
+    .with_context(|| format!("Failed to join on tenant {tenant_id} settings fetch task"))?
 }
 
-//
-// This is the entry point for the WAL receiver thread.
-//
-fn thread_main(conf: &'static PageServerConf, tenant_id: ZTenantId, timeline_id: ZTimelineId) {
-    let _enter = info_span!("WAL receiver", timeline = %timeline_id, tenant = %tenant_id).entered();
```

Comment:
This `info_span!` it pretty important for logging. It makes sure that for every error that happens when processing the incoming WAL has the timeline and tenant ID attached to it. I don't see anything to replace this.

Reply:
There's a bunch of others now, used as `.instrument(info_span` on a high-level futures such as `tokio::spawn` and such to replace this.

So far, the new span coverage looks good to me, here's an example: [test_pageserver_catchup logs](https://circleci-tasks-prod.s3.us-east-1.amazonaws.com/storage/artifacts/2d7b9e85-192b-4163-bc6c-303d0b8bb7d2/19954173-edff-40d2-9cce-53af749dd347/0/tmp/test_output/test_pageserver_catchup_while_compute_down/repo/pageserver.log?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=ASIAQVFQINEOEPKIPAUE%2F20220526%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20220526T084211Z&X-Amz-Expires=60&X-Amz-Security-Token=IQoJb3JpZ2luX2VjELn%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLWVhc3QtMSJHMEUCIQCyk41ON%2Bx5qCh4clO0BzuWSS6GC46UJgkkKNJQLs27mgIgLn0QL79D%2BDjxvGRZf%2FwGNGrKo6UPCbODmfZ7nXyNalIqtAIIov%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FARACGgwwNDU0NjY4MDY1NTYiDLciCOG8xPuNf6M4QiqIAqJBb%2FVVyHW%2Ft9qNk1AOQ%2Bz7DAmK0QrlDvxsZ26LvaPfaQc6Vku2fefSjtISC7qBowItoM9k%2FJFW%2BgNob76mgpyR4k50B8kyAfkSTkDFZTo3MpjN9Vxx67fwhCpYNAPTSVEhArW5eDbeNswxntTNS7SiC2mAKrmKOxQVMSgU0mRCroTk60yFRuD0rEuvOggjsLyZi0xu5n%2FgYTq3pqBBHq4G2eFFXfcuS8Mj9xcdkfER2Y%2FBpBG%2FA4D8cBTR6audm7n34S0vfETPZvaaJOIA8FcCeVBmdVuECIarYSjh0o3PyApZtDsiNKasP1Zq3p2A6YJhGvRFEdpgUg4PekX1zobVfXN0xskhGzDg%2BryUBjqdAZuJ0f0rti8vGxkHIrCoqUAzAvNagkV3NflqMK%2FJ4GMdP4MTvYTA5LOPDNqAIFww3dsB2EANy7o0DzUgZw2wVadOZYnCwk7Noeimr1yHe%2Fm8aBs4QpivGj%2BAeb2yiu%2FVMbqdwxyZYX3jJkWyjyjrp8J8k9%2ByLdr1zkOVl5ccTv2BNtYsH2eAaRjW60UGGtPSVB2vUoLKigwV1n21WS4%3D&X-Amz-SignedHeaders=host&x-id=GetObject&X-Amz-Signature=397172ba7521750c8a9d25a63ca13fff5927a37acf890b11b34c0aba626fc0ef)
(grep by `wal_receiver_main`)

Not sure about span names and data placement, but looks ok-ish for the first attempt too, IMO.